lib/Doctrine/DBAL/Query/Expression/CompositeExpression.php

```diff
@@ -43,47 +47,48 @@ class CompositeExpression implements Countable
      */
     public function __construct(string $type, array $parts = [])
     {
-        $this->type = $type;
-
-        $this->addMultiple($parts);
+        $this->type  = $type;
+        $this->parts = array_values(array_filter($parts, static function ($part) {
+            return ! ($part instanceof self && count($part) === 0);
+        }));
     }
 
     /**
-     * Adds multiple parts to composite expression.
+     * Returns a CompositeExpression with the given parts added.
      *
-     * @param array<int, self|string> $parts
+     * This instance is immutable and unaffected by this method call.
      *
-     * @return $this
+     * @param array<int, self|string> $parts
      */
     public function addMultiple(array $parts = []) : self
     {
+        $result = $this;
+
         foreach ($parts as $part) {
-            $this->add($part);
+            $result = $result->add($part);
         }
 
-        return $this;
+        return $result;
     }
 
     /**
-     * Adds an expression to composite expression.
+     * Returns a CompositeExpression with the given part added.
      *
-     * @param self|string $part
+     * This instance is immutable and unaffected by this method call.
      *
-     * @return $this
+     * @param self|string $part
      */
     public function add($part) : self
```

Comment:
Let’s rename it to `with()`:
1. A significant change in the method semantics without the signature change is quite dangerous.
2. The "with" naming is quite common for immutable class methods.

Reply:
If we have a new name, this could be contributed to 2.11.x as a new method (and we deprecate `add` and `addMultiple`)

Reply:
@morozov Agreed.
@greg0ire I don't think this is a good idea, as the semantics of the class differ. You'd end up with a mutable class that has mutator methods and non-mutator methods for the same thing, which would be highly confusing IMO.

Reply:
Wouldn't the confusion vanish on seeing that mutator methods are deprecated?

Reply:
IDK. @morozov?

Reply:
I believe we should introduce the new API in `2.11` w/o declaring the class itself immutable. This way, it will be at the consumer’s discretion to use one of the APIs before upgrading to the next major.

Reply:
Let's do this then. How should we proceed? Something along these lines?

1. retarget this PR on 2.11, just add `with()`
2. merge 2.11 into master
3. create another PR on master to remove mutators?

Reply:
Yes, correct. Go ahead.

Reply:
Done.